### ISSUE TITLE
a probable BUG in update policy when using GRPO

### ISSUE THREAD

**Issue by Racktic** [hiyouga/EasyR1#212]
In the function of [`update_policy`](https://github.com/hiyouga/EasyR1/blob/main/verl/workers/actor/dp_actor.py#L251), when computing `pg_loss`, we use [`VF.masked_mean`](https://github.com/hiyouga/EasyR1/blob/main/verl/trainer/core_algos.py#L349).
```
final_pg_loss = VF.masked_mean(final_pg_loss, response_mask)
```
And `VF.masked_mean` is defined as:
```python
def masked_mean(values: torch.Tensor, mask: torch.Tensor, dim: int = None, eps: float = 1e-8) -> torch.Tensor:
    """Compute mean of tensor with a masked values."""
    return (values * mask).sum(dim=dim) / (mask.sum(dim=dim) + eps)
```
Since `dim` is not specified, the `final_pg_loss` is compute across all tokens in this batch, which is actually the implementation of [**DAPO**](https://dapo-sia.github.io/static/pdf/dapo_paper.pdf).

![Image](https://github.com/user-attachments/assets/40b4aae9-4b5d-4a35-adae-eb8fcd359151)

However, **vanilla GRPO** in this [paper](http://arxiv.org/abs/2402.03300), is to compute average in one sequence first and then compute average across sequence.

![Image](https://github.com/user-attachments/assets/f29bc3cf-42cf-40c8-876b-07021729ca0c)

In a nutshell, current GRPO algorithm implementation seems is actually the implementation of DAPO🤔

**Comment by hiyouga** [comment]
Sure, it is what we initially intended to do. Sequence-level loss appears unstable in long-sequence tasks

**Comment by Racktic** [comment]
> Sure, it is what we initially intended to do. Sequence-level loss appears unstable in long-sequence tasks

Get it. Thanks for immediate reply : )
